src/MOI_wrapper.jl

```diff
@@ -118,604 +209,2102 @@ end
+    (MOI.SOLUTION_LIMIT, "Optimization terminated because the number of solutions found reached the value specified in the SolutionLimit parameter."),
+    (MOI.INTERRUPTED, "Optimization was terminated by the user."),
+    (MOI.NUMERICAL_ERROR, "Optimization was terminated due to unrecoverable numerical difficulties."),
+    (MOI.OTHER_LIMIT, "Unable to satisfy optimality tolerances; a sub-optimal solution is available."),
+    (MOI.OTHER_ERROR, "An asynchronous optimization call was made, but the associated optimization run is not yet complete."),
+    (MOI.OBJECTIVE_LIMIT, "User specified an objective limit (a bound on either the best objective or the best bound), and that limit has been reached.")
+]
+
+function MOI.get(model::Optimizer, ::MOI.RawStatusString)
+    status_code = get_status_code(model.inner)
+    if 1 <= status_code <= length(RAW_STATUS_STRINGS)
+        return RAW_STATUS_STRINGS[status_code][2]
+    end
+    return MOI.OTHER_ERROR
+end
+
+function MOI.get(model::Optimizer, ::MOI.TerminationStatus)
+    status_code = get_status_code(model.inner)
+    if 1 <= status_code <= length(RAW_STATUS_STRINGS)
+        return RAW_STATUS_STRINGS[status_code][1]
     end
     return MOI.OTHER_ERROR
 end
 
-function LQOI.get_primal_status(model::Optimizer)
+function MOI.get(model::Optimizer, ::MOI.PrimalStatus)
     stat = get_status(model.inner)
     if stat == :optimal
         return MOI.FEASIBLE_POINT
     elseif stat == :solution_limit
         return MOI.FEASIBLE_POINT
-    elseif (stat == :inf_or_unbd || stat == :unbounded) && has_primal_ray(model)
+    elseif (stat == :inf_or_unbd || stat == :unbounded) && _has_primal_ray(model)
         return MOI.INFEASIBILITY_CERTIFICATE
     elseif stat == :suboptimal
         return MOI.FEASIBLE_POINT
     elseif is_mip(model.inner) && get_sol_count(model.inner) > 0
         return MOI.FEASIBLE_POINT
-    else
-        return MOI.NO_SOLUTION
+    end
+    return MOI.NO_SOLUTION
+end
+
+function _has_dual_ray(model::Optimizer)
+    try
+        # Note: for performance reasons, we try to get 1 element because for
+        # some versions of Gurobi, we cannot query 0 elements without error.
+        get_dblattrarray(model.inner, "FarkasDual", 1, 1)
+        return true
+    catch ex
+        if isa(ex, GurobiError)
+            return false
+        else
+            rethrow(ex)
+        end
     end
 end
 
-function LQOI.get_dual_status(model::Optimizer)
+function MOI.get(model::Optimizer, ::MOI.DualStatus)
     stat = get_status(model.inner)
-    if is_mip(model.inner) || is_qcp(model.inner)
+    if is_mip(model.inner)
         return MOI.NO_SOLUTION
-    else
-        if stat == :optimal
-            return MOI.FEASIBLE_POINT
-        elseif stat == :solution_limit
-            return MOI.FEASIBLE_POINT
-        elseif (stat == :inf_or_unbd || stat == :infeasible) && has_dual_ray(model)
-            return MOI.INFEASIBILITY_CERTIFICATE
-        elseif stat == :suboptimal
-            r
```

Comment:
Why update if the variable is lower or upper bounded, but not if fixed or lower-and-upper bounded?

Reply:
We probably don't need to.